File: TBGTestingSD/wwwroot/js/Detail.js
var _id = $("#inputId").val(),
    _dataHeader,
    _dataDetail,
    _typeForm = 'add',
    _typeMaterial;
$(document).ready(function () {
    control.init();
    submit.init();
})

var control = {
    init: function () {
        control.getHeader(control.setInputDate);
        window.setInterval(control.bindExistData(), 500);
        $('#tglDiterima').datepicker({ format: 'yyyy-mm-dd' });
        $("#tipeMaterial").on('change', function () {
            var Type = $(this).val();
            control.getMaterial(Type);
        });
    },
    getHeader: function (callbackFn) {
        $.ajax({
            type: 'get',
            url: '/Header/GetById?id=' + _id,
            cache: false,
            success: function (obj, textStatus) {
                _dataHeader = obj;
                callbackFn();
            },
            error: function (obj, textStatus) {
                alert('error');
            }
        })
    },
    getMaterial: function (type) {
        if (type != '') {
            $.ajax({
                type: 'get',
                url: '/Detail/GetMaterial?tipe=' + type,
                cache: false,
                success: function (obj, textStatus) {
                    console.log(obj);
                    $('#material').text('');
                    $('#material').prop('disabled', false);
                    $('#material').prop('required', true);
                    $('#material').append('<option value="">pilih material..</option>');
                    var strHtml = '';
                    $.each(obj, function (k, v) {
                        strHtml = '<option value=' + v.id + '>' + v.material + '</option>'
                        $('#material').append(strHtml);
                    })
                },
                error: function (obj, textStatus) {
                    alert('error');
                }
            });
        } else {
            $('#material').text('');
            $('#material').prop('disabled', true);
            $('#material').prop('required', false);
        }
    },
    setInputDate: function () {
        var dateSPR = new Date(_dataHeader.tanggalMinta);
        dateSPR.setDate(dateSPR.getDate() + 14);
        console.log(dateSPR);
        console.log(new Date(dateSPR));
        var result = dateSPR.toLocaleString().split('T')[0];
        console.log(result);
        $('#tglDiterima').datepicker('setDate', new Date(result));
    },
    bindExistData: function () {
        $.ajax({
            type: 'get',
            url: '/Detail/GetDetailByIdHeader?id=' + _id,
            cache: false,
            success: function (obj, textStatus) {
                console.log(obj);
                if (obj) {
                    _typeForm = 'update';
                    _dataDetail = obj.detilSpr;
                    _typeMaterial = obj.tipeMaterial;
                    form.init();
                }
            },
            error: function (obj, textStatus) {
                alert('error');
            }
        })
    }
}

var form = {
    init: function () {
        $("#btnChange").off().on('click', function () {
            form.btnChange();
        });
        $("#btnCancel").off().on('click', function () {
            form.btnCancel();
        });
        form.btnCancel();  
    },
    update: function () {
        var obj = _dataDetail;
        $('#tipeMaterial').val((_typeMaterial) ? 1 : 0).change();
        $('#volume').val(obj.volume);
        $('#unit').val(obj.unit);
        $('#statusDisetujui').val((obj.statusDisetujui) ? 1 : 0).change();
        var date = new Date(obj.tanggalRencanaDiterima).toLocaleString();
        console.log(date);
        console.log(new Date(obj.tanggalRencanaDiterima).toLocaleString());
        $('#tglDiterima').datepicker('setDate', new Date(date));
        window.setTimeout(function () {
            $('#material').val(obj.material).change()
            $('.toogle').prop('disabled', true);
        },500)
    },
    btnChange: function () {
        $('.toogle').prop('disabled', false);
        $("#btnChange").prop('hidden', true);
        $("#btnBack").prop('hidden', true);
        $("#btnSave").prop('hidden', false);
        $("#btnCancel").prop('hidden', false);
    },
    btnCancel: function () {
        form.update();
        $("#btnChange").prop('hidden', false);
        $("#btnBack").prop('hidden', false);
        $("#btnSave").prop('hidden', true);
        $("#btnCancel").prop('hidden', true);
    },
}

var submit = {
    init: function () {
        $("#formDetail").submit(function (e) {
            e.preventDefault();
            if (_typeForm == 'add') {
                submit.add();
            } else {
                submit.update();
            }
        })
    },
    add: function () {
        var params = new Object();
        params.IdRef = _id;
        params.Material = $("#material").val();
        params.Volume = $("#volume").val();
        params.Unit = $("#unit").val();
        params.StatusDisetujui = ($("#statusDisetujui").val() == 1);
        params.TanggalRencanaDiterima = $("#tglDiterima").val();

        $.ajax({
            type: 'post',
            url: '/Detail/AddNew',
            data: params,
            cache: false,
            success(data) {
                if (data == 'pass') {
                    alert('success');
                    window.location.reload();
                } else {
                    alert(data.message);
                }
            },
            error(data) {
                alert('error')
            }
        })
    },
    update: function () {
        var params = new Object();
        params.Id = _dataDetail.id;
        params.IdRef = _id;
        params.Material = $("#material").val();
        params.Volume = $("#volume").val();
        params.Unit = $("#unit").val();
        params.StatusDisetujui = ($("#StatusDisetujui").val() == 1);
        params.TanggalRencanaDiterima = $("#tglDiterima").val();

        $.ajax({
            type: 'post',
            url: '/Detail/Update',
            data: params,
            cache: false,
            success(data) {
                if (data == 'pass') {
                    alert('success');
                    window.location.reload();
                } else {
                    alert(data.message);
                }
            },
            error() {
                alert('error')
            }
        })
    }
}

File: RepositoryLayer/Service/Detail.cs
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using RepositoryLayer.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace RepositoryLayer.Service
{
    public class Detail
    {
        private readonly TBPContext context;
        public Detail(TBPContext TBPContext)
        {
            context = TBPContext;
        }
        public DetilSpr GetById(int id)
        {
            return context.DetilSpr.FirstOrDefault(o => o.Id == id);
        }
        public DetilSpr GetByIdRef(int idRef)
        {
            return context.DetilSpr.FirstOrDefault(o => o.IdRef == idRef);
        }
        public string Remove(int id)
        {
            string result = string.Empty;
            var data = context.DetilSpr.FirstOrDefault(o => o.Id == id);
            if (data != null)
            {
                context.DetilSpr.Remove(data);
                result = "pass";
            }
            return result;
        }
        public string AddNew(DetilSpr obj)
        {
            try
            {
                string result = string.Empty;

                var newDetail = new DetilSpr();
                newDetail.IdRef = obj.IdRef;
                newDetail.Material = obj.Material;
                newDetail.Unit = obj.Unit;
                newDetail.Volume = obj.Volume;
                newDetail.StatusDisetujui = obj.StatusDisetujui;
                newDetail.TanggalRencanaDiterima = obj.TanggalRencanaDiterima;

                var updatedHeader = context.HeaderSpr.FirstOrDefault(c => c.Id == obj.IdRef);
                updatedHeader.AdaDetil = true;

                context.DetilSpr.Add(newDetail);
                context.SaveChanges();

                result = "pass";

                return result;
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }

        public string Update(DetilSpr obj)
        {
            try
            {
                var existedData = context.DetilSpr.FirstOrDefault(c => c.Id == obj.Id);

                existedData.StatusDisetujui = obj.StatusDisetujui;
                existedData.Unit = obj.Unit;
                existedData.Volume = obj.Volume;
                existedData.Material = obj.Material;
                existedData.TanggalRencanaDiterima = obj.TanggalRencanaDiterima;

                context.SaveChanges();

                string result  = "pass";
                return result;
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
        }
    }
}
